File: sol_auto_transfer_env.py
import os
import time
import random
import base58
from dotenv import load_dotenv
from solana.rpc.api import Client
from solders.keypair import Keypair
from solders.pubkey import Pubkey
from solders.transaction import Transaction
from solders.system_program import TransferParams, transfer

LAMPORTS_PER_SOL = 1_000_000_000

# ========= 读取 .env =========
load_dotenv()

KEY_A = os.getenv("SOL_WALLET_A_PRIVATE_KEY")
KEY_B = os.getenv("SOL_WALLET_B_PRIVATE_KEY")
RPC_URL = os.getenv("SOL_RPC_URL", "https://api.mainnet-beta.solana.com")

if not KEY_A or not KEY_B:
    raise RuntimeError("❌ .env 中缺少钱包私钥")

client = Client(RPC_URL)

# ========= 配置 =========

INTERVAL_SECONDS = 10 * 60
MIN_SOL = 0.005
MAX_SOL = 0.01

# ========= 钱包初始化 =========

def keypair_from_base58(b58_key: str) -> Keypair:
    secret = base58.b58decode(b58_key)
    return Keypair.from_bytes(secret)

wallet_a = keypair_from_base58(KEY_A)
wallet_b = keypair_from_base58(KEY_B)

# ========= 工具函数 =========

def random_amount_lamports():
    sol = round(random.uniform(MIN_SOL, MAX_SOL), 6)
    return sol, int(sol * LAMPORTS_PER_SOL)

def send_sol(sender: Keypair, receiver: Pubkey, lamports: int) -> str:
    ix = transfer(
        TransferParams(
            from_pubkey=sender.pubkey(),
            to_pubkey=receiver,
            lamports=lamports,
        )
    )
    
    # 获取最新区块哈希
    blockhash = client.get_latest_blockhash().value.blockhash
    
    # 构建并签名交易
    tx = Transaction.new_signed_with_payer(
        [ix],
        sender.pubkey(),
        [sender],
        blockhash
    )

    # 发送交易
    resp = client.send_transaction(tx)
    
    # 发送失败通常会抛出异常，这里主要处理 RPC 返回结构
    # solana-py 0.36+ 的 resp.value 是 Signature 对象
    if not resp.value:
         raise Exception(f"RPC异常: {resp}")

    sig = str(resp.value)
    
    # 确认交易 (可选，为了保持原逻辑一致性)
    client.confirm_transaction(resp.value, commitment="confirmed")
    return sig

# ========= 主循环 =========

def main():
    direction = True  # True: A->B, False: B->A

    print("🔐 私钥来源：.env 文件")
    print("🚀 自动转账启动（0.005 ~ 0.01 SOL，每10分钟）")
    print("A:", wallet_a.pubkey())
    print("B:", wallet_b.pubkey())
    print("RPC:", RPC_URL)

    while True:
        try:
            sol, lamports = random_amount_lamports()

            if direction:
                sig = send_sol(wallet_a, wallet_b.pubkey(), lamports)
                print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] A → B | {sol} SOL | {sig}")
            else:
                sig = send_sol(wallet_b, wallet_a.pubkey(), lamports)
                print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] B → A | {sol} SOL | {sig}")

            direction = not direction

        except Exception as e:
            print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] ❌ 转账失败: {e}")

        time.sleep(INTERVAL_SECONDS)

if __name__ == "__main__":
    main()

File: sol_auto_transfer_env.js
import "dotenv/config";
import bs58 from "bs58";
import {
    Connection,
    Keypair,
    SystemProgram,
    Transaction,
    sendAndConfirmTransaction,
    LAMPORTS_PER_SOL,
} from "@solana/web3.js";

// ======== 配置 ========

const RPC_URL = process.env.SOL_RPC_URL || "https://api.mainnet-beta.solana.com";
const INTERVAL_MS = 10 * 60 * 1000;

const MIN_SOL = 0.005;
const MAX_SOL = 0.01;

// ======== 从 .env 读取私钥 ========

const KEY_A = process.env.SOL_WALLET_A_PRIVATE_KEY;
const KEY_B = process.env.SOL_WALLET_B_PRIVATE_KEY;

if (!KEY_A || !KEY_B) {
    throw new Error("❌ .env 缺少 SOL_WALLET_A_PRIVATE_KEY 或 SOL_WALLET_B_PRIVATE_KEY");
}

function keypairFromBase58(b58) {
    const secretKey = bs58.decode(b58.trim());
    return Keypair.fromSecretKey(secretKey);
}

const walletA = keypairFromBase58(KEY_A);
const walletB = keypairFromBase58(KEY_B);

const connection = new Connection(RPC_URL, "confirmed");

// ======== 工具函数 ========

function randomAmountLamports() {
    // 保留6位小数，避免太多尾数
    const sol = Number((Math.random() * (MAX_SOL - MIN_SOL) + MIN_SOL).toFixed(6));
    const lamports = BigInt(Math.round(sol * LAMPORTS_PER_SOL));
    return { sol, lamports };
}

async function transferSOL(from, toPubkey, lamportsBigInt) {
    // web3.js 这里接受 number；但 lamports 可能超过安全整数？本场景很小，安全。
    const lamports = Number(lamportsBigInt);

    const tx = new Transaction().add(
        SystemProgram.transfer({
            fromPubkey: from.publicKey,
            toPubkey: toPubkey,
            lamports,
        })
    );

    const sig = await sendAndConfirmTransaction(connection, tx, [from]);
    return sig;
}

// ======== 主循环 ========

let direction = true; // true: A->B, false: B->A

async function loopOnce() {
    const { sol, lamports } = randomAmountLamports();

    try {
        if (direction) {
            const sig = await transferSOL(walletA, walletB.publicKey, lamports);
            console.log(`[${new Date().toLocaleString()}] A → B | ${sol} SOL | tx: ${sig}`);
        } else {
            const sig = await transferSOL(walletB, walletA.publicKey, lamports);
            console.log(`[${new Date().toLocaleString()}] B → A | ${sol} SOL | tx: ${sig}`);
        }

        direction = !direction;
    } catch (e) {
        console.error(`[${new Date().toLocaleString()}] ❌ 转账失败:`, e?.message || e);
        // 失败时不切换方向：保持下次仍然尝试同方向（更稳）
        // direction 不变
    }
}

// 立即跑一次
await loopOnce();

// 每10分钟跑一次
setInterval(() => {
    loopOnce().catch((e) => console.error("loop error:", e?.message || e));
}, INTERVAL_MS);

console.log("✅ 启动成功");
console.log("RPC:", RPC_URL);
console.log("A:", walletA.publicKey.toBase58());
console.log("B:", walletB.publicKey.toBase58());
console.log(`规则：每 ${INTERVAL_MS / 60000} 分钟，随机转 ${MIN_SOL}～${MAX_SOL} SOL，方向轮流`);
